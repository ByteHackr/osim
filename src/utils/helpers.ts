import { toRaw, isRef, isReactive, isProxy, ref, toRef, watch } from 'vue';
import * as R from 'ramda';

export function vRef(prop: Record<string, any>, property: string, defaultValue: any) {
  const reffedProp = toRef(prop, property);
  watch(reffedProp, (value) => flexRef.value = value);
  const flexRef = reffedProp.value === undefined ? ref(defaultValue) : reffedProp;
  return flexRef;
}

export function deepToRaw<T extends Record<string, any>>(sourceObj: T): T {
  const objectIterator = (input: any): any => {
    if (Array.isArray(input)) {
      return input.map((item) => objectIterator(item));
    }
    if (isRef(input) || isReactive(input) || isProxy(input)) {
      return objectIterator(toRaw(input));
    }
    if (input && typeof input === 'object') {
      return Object.keys(input).reduce((acc, key) => {
        acc[key as keyof typeof acc] = objectIterator(input[key]);
        return acc;
      }, {} as T);
    }
    return input;
  };

  return objectIterator(sourceObj);
}

export function deepCopyFromRaw<T extends Record<string, any>>(sourceObj: T): T {
  return JSON.parse(JSON.stringify(deepToRaw(sourceObj)));
}

// eslint-disable-next-line max-len
// https://stackoverflow.com/questions/14446511/most-efficient-method-to-groupby-on-an-array-of-objects
export const groupBy = <T>(
  array: T[],
  predicate: (value: T, index: number, array: T[]) => string,
) =>
    array.reduce(
      (acc, value, index, array) => {
        (acc[predicate(value, index, array)] ||= []).push(value);
        return acc;
      },
    {} as { [key: string]: T[] },
    );

export const assignKeyValue = (object: Record<string, any>, key: string, value: any = null) => {
  object[key] = value;
  return object;
};

export const objectMap = (object: Record<string, any>, mapFn: (key: string, value: any) => any) =>
  Object.keys(object).reduce(
    (acc, key) => assignKeyValue(acc, key, mapFn(key, object[key])),
    {} as Record<string, any>,
  );

export const sortedByGroup = <T extends Record<string, any>>(array: T[], key: string) =>
  groupBy(
    array
      .filter((item: T) => item[key])
      .sort((itemA: T, itemB: T) => itemA[key].localeCompare(itemB[key])),
    (item: T) => item[key],
  );

type DeepMappable = any[] | Record<string, any>

const isNonEmptyArray = (value: any) => R.is(Array, value) && value.length > 0;
const isNonArrayObject = (value: any) => R.is(Object, value) && !R.is(Array, value);
const isDeepMappable = (value: DeepMappable) => isNonEmptyArray(value) || isNonArrayObject(value);

export const deepMap = (transform: (arg: any) => any, object: DeepMappable): any =>
  R.map(
    (val: any) => isDeepMappable(val)
      ? deepMap(transform, val)
      : transform(val)
    , object
  );

export const cveRegex = /^\s*(CVE-\d{4}-\d{4,7})\s*$/;
