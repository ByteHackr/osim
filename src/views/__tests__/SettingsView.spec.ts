import { mountWithConfig } from '@/__tests__/helpers';

import SettingsView from '../SettingsView.vue';

describe('settingsView', () => {
  it('should render', () => {
    const wrapper = mountWithConfig(SettingsView);

    expect(wrapper.html()).toMatchSnapshot();
  });
});
