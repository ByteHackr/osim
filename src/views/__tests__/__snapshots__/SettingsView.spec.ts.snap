// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`settingsView > should render 1`] = `
"<main class="mt-3">
  <div data-v-511ff0d3="" class="osim-content container">
    <h1 data-v-511ff0d3="" class="mb-3">Settings</h1>
    <div data-v-511ff0d3="" class="alert alert-info" role="alert"> These values are saved in browser local storage, and should persist across tabs and browser sessions. Ensure that security best practices are followed. </div>
    <!--@submit.prevent="settingsStore.save(settings)"-->
    <form data-v-511ff0d3="" class="osim-settings">
      <!-- <LabelInput
        v-model="bugzillaApiKey"
        label="Bugzilla API Key"
        :error="errors.bugzillaApiKey"
      /> -->
      <div data-v-511ff0d3="" class="form-control mb-3"><label data-v-511ff0d3="" class="form-check"><input data-v-511ff0d3="" class="form-check-input" type="radio" name="revealSensitive" value="password"><span data-v-511ff0d3="" class="form-check-label">Hide Password Values</span></label><label data-v-511ff0d3="" class="form-check"><input data-v-511ff0d3="" class="form-check-input" type="radio" name="revealSensitive" value="text"><span data-v-511ff0d3="" class="form-check-label">Reveal Password Values</span></label></div>
      <div data-v-511ff0d3="" class="form-control mb-3"><label data-v-511ff0d3="" class="d-block"><span data-v-511ff0d3="" class="form-label">Bugzilla API Key</span><input data-v-511ff0d3="" class="form-control is-invalid" type="password" placeholder="[none saved]"><span data-v-511ff0d3="" class="invalid-feedback d-block"> Please provide a Bugzilla key. </span></label>
        <div data-v-511ff0d3="" class="form-text">
          <p data-v-511ff0d3="">Required for actions which interface with Bugzilla.</p>
          <details data-v-511ff0d3="">
            <summary data-v-511ff0d3="">Steps to create an API key:</summary>
            <ul data-v-511ff0d3="">
              <li data-v-511ff0d3="">Visit your Bugzilla user preferences page</li>
              <li data-v-511ff0d3="">Click the API Keys tab</li>
              <li data-v-511ff0d3="">Check the "Generate a new API key" box</li>
              <li data-v-511ff0d3="">Click the "Submit Changes" button</li>
              <li data-v-511ff0d3=""> Copy the API key from the banner at the top of the page. Optionally, save it to your password manager. The API key will not be shown again after you navigate away from the page. </li>
            </ul>
          </details>
        </div>
      </div>
      <div data-v-511ff0d3="" class="form-control mb-3"><label data-v-511ff0d3="" class="d-block"><span data-v-511ff0d3="" class="form-label">JIRA API Key</span><input data-v-511ff0d3="" class="form-control is-invalid" type="password" placeholder="[none saved]"><span data-v-511ff0d3="" class="invalid-feedback d-block"> Please provide a Jira key. </span></label>
        <div data-v-511ff0d3="" class="form-text">
          <p data-v-511ff0d3="">Required for actions which interface with JIRA.</p>
          <details data-v-511ff0d3="">
            <summary data-v-511ff0d3="">Steps to create an API key:</summary>
            <ul data-v-511ff0d3="">
              <li data-v-511ff0d3="">Log into your JIRA account.</li>
              <li data-v-511ff0d3=""> Go to 'Your Profile' page under your user icon menu in the upper right corner. </li>
              <li data-v-511ff0d3="">Select Personal Access Tokens from the left hand sidebar menu.</li>
              <li data-v-511ff0d3="">Click Create token button.</li>
              <li data-v-511ff0d3="">Name the token meaningfully (eg. 'OSIM token') and click create.</li>
              <li data-v-511ff0d3="">Copy token value and close.</li>
            </ul>
          </details>
        </div>
      </div>
    </form>
  </div>
</main>"
`;
